File: src/main/org/nlogo/hubnet/mirroring/Overridable.scala
// (C) 2012 Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.hubnet.mirroring

import java.lang.reflect.{ InvocationTargetException, Method }

object Overridable {
  def getOverrideIndex(variables: Array[String], varName: String): Int =
    variables.indexWhere(_ equalsIgnoreCase varName)
  def getOverrideIndex(agentType: AgentType, varName: String) =
    agentType match {
      case AgentType.Turtle =>
        TurtleData.getOverrideIndex(varName)
      case AgentType.Patch =>
        PatchData.getOverrideIndex(varName)
      case AgentType.Link =>
        LinkData.getOverrideIndex(varName)
      case AgentType.Observer =>
        throw new IllegalArgumentException
    }
}

abstract class Overridable {

  private var stack: List[(Method, AnyRef)] = Nil

  def getterName(varName: Int): String  // abstract
  def setterName(varName: Int): String  // abstract

  def set(variable: Int, value: Any) {
    val gName = getterName(variable)
    val getter = getClass.getMethod(gName)
    val sName = setterName(variable)
    val setter = getSetter(sName, value.getClass)
    // note that the setter for the old value might be different
    // than the setter for the new value ev 4/29/08
    val oldValue = getter.invoke(this)
    stack ::= (getSetter(sName, oldValue.getClass), oldValue)
    setter.invoke(this, value.asInstanceOf[AnyRef])
  }

  private def getSetter(methodName: String, clazz: Class[_]): Method = {
    // when a value type is passed as a parameter to a method that takes Any, it gets boxed, so
    // when we call getClass above, we get a boxed class.  but the setter method we're looking
    // for is unboxed, so when we do method lookup we need to substitute a Class object
    // representing the (not actually existing) unboxed class - ST 1/21/12
    val unboxedClass =
      if(clazz eq classOf[java.lang.Double])
        classOf[Double]
      else if(clazz eq classOf[java.lang.Integer])
        classOf[Int]
      else if(clazz eq classOf[java.lang.Boolean])
        classOf[Boolean]
      else
        clazz
    try getClass.getMethod(methodName, unboxedClass)
    catch {
      case ex: NoSuchMethodException =>
        // if we don't find a setter specific for this type look for an Object type setter ev 5/16/08
        getClass.getMethod(methodName, classOf[AnyRef])
    }
  }

  def rollback() {
    for((method, value) <- stack)
      method.invoke(this, value)
    stack = Nil
  }

}

File: headless/src/main/org/nlogo/hubnet/mirroring/Overridable.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.hubnet.mirroring;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public abstract strictfp class Overridable {
  private final RollbackStack rollbackStack = new RollbackStack();

  abstract String getMethodName(int varName);

  public void set(int var, Object value) {
    try {
      String methodName = getMethodName(var);
      Method getter = getClass().getMethod(methodName);
      Method setter = getSetter(methodName, value.getClass());
      // note that the setter for the old value might be different
      // than the setter for the new value ev 4/29/08
      Object oldValue = getter.invoke(this);
      rollbackStack.push(getSetter(methodName, oldValue.getClass()), oldValue);
      setter.invoke(this, value);
    } catch (NoSuchMethodException ex) {
      throw new IllegalStateException(ex);
    } catch (IllegalAccessException e) {
      throw new IllegalStateException(e);
    } catch (InvocationTargetException e) {
      throw new IllegalStateException(e);
    }
  }

  private Method getSetter(String methodName, Class<? extends Object> class1)
      throws NoSuchMethodException {
    try {
      return getClass().getMethod(methodName, class1);
    } catch (NoSuchMethodException ex) {
      // if we don't find a setter specific for this type look for an Object
      // type setter ev 5/16/08
      return getClass().getMethod(methodName, Object.class);
    }
  }

  public static int getOverrideIndex(String[] variables, String varName) {
    for (int i = 0; i < variables.length; i++) {
      if (varName.equalsIgnoreCase(variables[i])) {
        return i;
      }
    }
    return -1;
  }

  public void rollback() {
    rollbackStack.rollback(this);
  }

  private strictfp class RollbackStack {
    private final java.util.LinkedList<Method> methodStack = new java.util.LinkedList<Method>();
    private final java.util.LinkedList<Object> valueStack = new java.util.LinkedList<Object>();

    void push(Method setter, Object value) {
      methodStack.addLast(setter);
      valueStack.addLast(value);
    }

    void rollback(Object owner) {
      try {
        while (!methodStack.isEmpty()) {
          methodStack.removeLast().invoke(owner, valueStack.removeLast());
        }
      } catch (IllegalAccessException e) {
        throw new IllegalStateException(e);
      } catch (InvocationTargetException e) {
        throw new IllegalStateException(e);
      }
    }
  }
}
